### PR TITLE
BUG: Assorted DTA/TDA/PI bugfixes

### LINKED ISSUE #23078: comparison between two PeriodIndexes doesn't validate length
```python
In [5]: idx = pd.period_range('2017', periods=12, freq="A-DEC")

In [6]: idx <= idx[[0]]
Out[6]:
array([ True, False, False, False, False, False, False, False, False,
       False, False, False])

```

This should raise.

Both `DatetimeIndex` and `TimedeltaIndex` do raise

```python
In [8]: idx <= idx[[0]]
```

```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-8-8a2a887e6efd> in <module>
----> 1 idx <= idx[[0]]

~/sandbox/pandas/pandas/core/indexes/datetimes.py in wrapper(self, other)
     90
     91     def wrapper(self, other):
---> 92         result = getattr(DatetimeArrayMixin, opname)(self, other)
     93         if is_bool_dtype(result):
     94             return result

~/sandbox/pandas/pandas/core/arrays/datetimes.py in wrapper(self, other)
    133             else:
    134                 self._assert_tzawareness_compat(other)
--> 135                 result = meth(self, np.asarray(other))
    136
    137             result = com.values_from_object(result)

~/sandbox/pandas/pandas/core/arrays/datetimelike.py in cmp_method(self, other)
     51         if isinstance(other, (np.ndarray, ABCIndexClass, ABCSeries)):
     52             if other.ndim > 0 and len(self) != len(other):
---> 53                 raise ValueError('Lengths must match to compare')
     54
     55         if needs_i8_conversion(self) and needs_i8_conversion(other):

ValueError: Lengths must match to compare

```

### PR DESCRIPTION
**Pull request by jbrockmendel**
Fixes a problem with tz-aware DatetimeArray.round (ref #24038)
Fixes PI, PA, DTA failing to validate lengths when doing comparisons against length-1 instances (ref #23078)
Pins names to a handful of methods that currently don't have them (working on a separate branch that checks these systematically; these are the low-hanging fruit)

- [x] closes #23078
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/datetimes.py [hunk 1]
```diff
@@ -110,9 +112,12 @@ def wrapper(self, other):
             result = op(self.asi8, other.view('i8'))
             if isna(other):
                 result.fill(nat_result)
-        elif lib.is_scalar(other):
+        elif lib.is_scalar(other) or np.ndim(other) == 0:
             return ops.invalid_comparison(self, other, op)
         else:
+            if len(other) != len(self):
```
Comment (jreback) [inline pandas/core/arrays/datetimes.py]: should be part of the elif

File: pandas/core/arrays/timedeltas.py [hunk 2]
```diff
@@ -90,6 +90,9 @@ def wrapper(self, other):
             return ops.invalid_comparison(self, other, op)
 
         else:
+            if len(other) != len(self):
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: make a part of the elif

**Review by jreback — CHANGES_REQUESTED** [review]
minor code comments. can you add whatsnews as needed.

**Comment by jreback** [comment]
is this orthogonal to #23078, should close that one?

**Comment by jbrockmendel** [comment]
> is this orthogonal to #23078, should close that one?

This closes #23078, should be reflected in the OP.

**Comment by jbrockmendel** [comment]
i think comments have been addressed

File: pandas/core/ops.py [hunk 3]
```diff
@@ -1762,6 +1764,7 @@ def wrapper(self, other, axis=None):
             return self._constructor(res_values, index=self.index,
                                      name=res_name, dtype='bool')
 
+    wrapper.__name__ = op_name
```
Comment (jreback) [inline pandas/core/ops.py]: is somthing testing these?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/ops.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: No.  I have a local branch that checks these for all PandasObject subclasses, but there are a bunch of non-obvious cases in there.  This just fixes the more obvious cases.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks. pls followup on the ops name tests at some point.

**Pull request merged**